src/olympia/shelves/serializers.py

```diff
@@ -30,3 +35,14 @@ def get_url(self, obj):
             url = None
 
         return url
+
+    def get_addons(self, obj):
+        if obj.endpoint == 'search':
+            criteria = obj.criteria.strip('?')
+            params = dict(parse.parse_qsl(criteria))
+            request = self.context.get('request', None)
+            request.GET = request.GET.copy()
+            request.GET.update(params)
+            return AddonSearchView(request=request).data
```

Comment:
+ @bobsilverberg 
I got it to return the add-ons based on `criteria`, but I don't believe I did it correctly. `tests/test_serializers.py` fails with `AttributeError: 'WSGIRequest' object has no attribute 'query_params'`. When `query_params` is set, it goes on to fail with AttributeError: 'WSGIRequest' object has no attribute 'user'`.

Reply:
Somewhere along the way `request` has been unwrapped - DRF has it's own `Request` [1] object that wraps django's `HttpRequest` (`WSGIRequest`) and adds among other things, `query_params` and `user`.  I can't see the problem immediately so I'll have to have a look around.

[1] https://www.django-rest-framework.org/api-guide/requests/#query_params

Reply:
Oh, so `APIRequestFactory` that DRF provides doesn't actually return a DRF `Request` - it generates a normal django request.  Functional tests in test_views.py should work (the view creates the DRF `Request`).  We'll probably have to something similar to what `initialize_request` does for the serializer tests - maybe just `self.request = Request(APIRequestFactory().get('/api/%s/' % api_version))` is enough?

[1] https://github.com/encode/django-rest-framework/blob/3.11.1/rest_framework/views.py#L381

Reply:
The request is currently `<WSGIRequest: GET '/api/v4/'>`, but should the request that goes into `AddonSearchView(request=request).data` look like `<WSGIRequest: GET '/api/v4/search/?sort=users&type=extension'>`?

Reply:
mmm, good question.  I don't _think_ it matters - iirc the path to the api isn't checked within the view itself, just the version at the start.  But it's something to play around with if it doesn't work as expected (especially in tests)